pkg/keystone/keystone.go

```diff
@@ -384,6 +405,13 @@ func (d *keystone) authOptionsFromRequest(r *http.Request, guessScope bool) (*go
 
 		// set password
 		ba.Password = password
+		// if application credentials are used, skip th basic auth checks below
+	} else if (appcredid != "" && appcredsecret != "") ||
+		(appcredname != "" && appcredusername != "") {
+		ba.ApplicationCredentialID = appcredid
+		ba.ApplicationCredentialName = appcredname
```

Comment:
Yes.  appCredUserName is equal to username.  This gets set on line 354.  Since we're not using basic authentication, the username must be parsed from the headers, outside off the BasicAuth() call via the 'X-User-Name' header.